File: python/src/multiagent_boilerplate/tools/builtin/read_file.py
from __future__ import annotations

from pathlib import Path

from ..types import ToolContext

WORKSPACE_ROOT = Path.cwd().resolve()


class ReadFileTool:
    name = "read_file"
    description = "Read a UTF-8 text file by path, relative to the project working directory."
    input_schema = {"type": "object", "required": ["path"], "properties": {"path": {"type": "string"}}}
    exposable = False

    async def execute(self, input_: dict, ctx: ToolContext) -> dict:
        target = Path(input_["path"])
        target = target if target.is_absolute() else (WORKSPACE_ROOT / target)
        target = target.resolve()
        if not str(target).startswith(str(WORKSPACE_ROOT)):
            raise ValueError(f"read_file: path escapes workspace root: {input_['path']}")
        return {"content": target.read_text(encoding="utf-8")}


read_file_tool = ReadFileTool()

File: typescript/src/tools/builtin/readFile.ts
import { readFile } from "node:fs/promises";
import { resolve, isAbsolute } from "node:path";
import type { Tool } from "../types.js";

const workspaceRoot = process.cwd();

/** Reads a file scoped to the current working directory - agents cannot read outside it. */
export const readFileTool: Tool<{ path: string }, { content: string }> = {
  name: "read_file",
  description: "Read a UTF-8 text file by path, relative to the project working directory.",
  inputSchema: {
    type: "object",
    required: ["path"],
    properties: { path: { type: "string" } },
  },
  async execute(input) {
    const target = isAbsolute(input.path) ? input.path : resolve(workspaceRoot, input.path);
    if (!target.startsWith(workspaceRoot)) {
      throw new Error(`read_file: path escapes workspace root: ${input.path}`);
    }
    const content = await readFile(target, "utf-8");
    return { content };
  },
};
